flaskbb/utils/requirements.py

```diff
@@ -113,185 +122,217 @@ def _determine_locked(self, request):
+        IsModeratorInForum(forum_id=forum_id, forum=forum),
+    )
 
 
-IsMod = And(IsAuthed(), Has('mod'))
-IsSuperMod = And(IsAuthed(), Has('super_mod'))
-IsAdmin = And(IsAuthed(), Has('admin'))
+IsMod = And(IsAuthed(), Has("mod"))
+IsSuperMod = And(IsAuthed(), Has("super_mod"))
+IsAdmin = And(IsAuthed(), Has("admin"))
 
 IsAtleastModerator = Or(IsAdmin, IsSuperMod, IsMod)
 
 IsAtleastSuperModerator = Or(IsAdmin, IsSuperMod)
 
-CanBanUser = Or(IsAtleastSuperModerator, Has('mod_banuser'))
+CanBanUser = Or(IsAtleastSuperModerator, Has("mod_banuser"))
 
-CanEditUser = Or(IsAtleastSuperModerator, Has('mod_edituser'))
+CanEditUser = Or(IsAtleastSuperModerator, Has("mod_edituser"))
 
-CanEditPost = Or(IsAtleastSuperModerator,
-                 And(IsModeratorInForum(), Has('editpost')),
-                 And(IsSameUser(), Has('editpost'), TopicNotLocked()))
+CanEditPost = Or(
+    IsAtleastSuperModerator,
+    And(IsModeratorInForum(), Has("editpost")),
+    And(CanAccessForum(), IsSameUser(), Has("editpost"), TopicNotLocked()),
+)
 
 CanDeletePost = CanEditPost
 
-CanPostReply = Or(And(Has('postreply'), TopicNotLocked()),
-                  IsModeratorInForum(),
-                  IsAtleastSuperModerator)
+CanPostReply = Or(
+    And(Has("postreply"), TopicNotLocked()),
+    IsModeratorInForum(),
+    IsAtleastSuperModerator,
+)
 
-CanPostTopic = Or(And(Has('posttopic'), ForumNotLocked()),
-                  IsAtleastSuperModerator,
-                  IsModeratorInForum())
+CanPostTopic = Or(
+    And(Has("posttopic"), ForumNotLocked()),
+    IsAtleastSuperModerator,
+    IsModeratorInForum(),
+)
 
-CanDeleteTopic = Or(IsAtleastSuperModerator,
-                    And(IsModeratorInForum(), Has('deletetopic')),
-                    And(IsSameUser(), Has('deletetopic'), TopicNotLocked()))
+CanDeleteTopic = Or(
+    IsAtleastSuperModerator,
+    And(IsModeratorInForum(), Has("deletetopic")),
+    And(IsSameUser(), Has("deletetopic"), TopicNotLocked()),
+)
 
 
-# Template Allowances -- gross, I know
+def permission_with_identity(requirement, name=None):
+    """
+    Permission instance factory that can set a user at construction time
+    can optionally name the closure for nicer debugging
+    """
 
-def TplCanModerate(request):
-    def _(user, forum):
-        kwargs = {}
+    def _(user):
+        return Permission(requirement, identity=user)
 
-        if isinstance(forum, int):
-            kwargs['forum_id'] = forum
-        elif isinstance(forum, Forum):
-            kwargs['forum'] = forum
+    if name is not None:
+        _.__name__ = name
 
-        return IsAtleastModeratorInForum(**kwargs)(user, request)
     return _
 
 
-def TplCanPostReply(request):
-    def _(user, topic=None):
-        kwargs = {}
+# Template Requirements
+
 
-        if isinstance(topic, int):
-            kwargs['topic_id'] = topic
-        elif isinstance(topic, Topic):
-            kwargs['topic'] = topic
+def has_permission(permission):
```

Comment:
Basically a convenience function optimized for specific permission checking. We should probably use this over explicit permission checking to take advantage of things like override and additional requirements in 0.6.0 which can let us optionally bypass or strengthen some checks for users.